src/Dictionaries/HashedDictionary.cpp

```diff
@@ -35,16 +42,156 @@ namespace ErrorCodes
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            shards_queues[shard].emplace(backlog);
+            pool.scheduleOrThrowOnError([this, shard, thread_group = CurrentThread::getGroup()]
+            {
+                if (thread_group)
+                    CurrentThread::attachToIfDetached(thread_group);
+                setThreadName("HashedDictLoad");
+
+                threadWorker(shard);
+            });
+        }
+    }
+
+    void addBlock(Block block)
+    {
+        Blocks shards_blocks(shards);
+        for (size_t shard = 0; shard < shards; ++shard)
+            shards_blocks[shard] = block.cloneEmpty();
+
+        IColumn::Selector selector;
+        if (simple_key)
+            selector = createBlockSelector<UInt64>(*block.getByPosition(0).column, shards_slots);
+        else
+            selector = createComplexBlockSelector(block, shards_slots);
+
+        splitBlock(selector, block, shards_blocks);
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            if (!shards_queues[shard]->push(std::move(shards_blocks[shard])))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to shards queue #{}", shard);
+        }
+    }
+
+    void finish()
+    {
+        for (auto & queue : shards_queues)
+            queue->finish();
+
+        Stopwatch watch;
+        pool.wait();
+        UInt64 elapsed_us = watch.elapsedMicroseconds();
+        LOG_TRACE(dictionary.log, "Processing the tail took {}ms", elapsed_us);
+    }
+
+    ~ParallelDictionaryLoader()
+    {
+        for (auto & queue : shards_queues)
+            queue->clearAndFinish();
+        pool.wait();
+    }
+
+private:
+    HashedDictionary & dictionary;
+    const size_t shards;
+    bool simple_key;
+    ThreadPool pool;
+    std::vector<std::optional<Queue>> shards_queues;
+    std::vector<UInt64> shards_slots;
+
+    void threadWorker(size_t shard)
+    {
+        Block block;
+        auto & shard_queue = *shards_queues[shard];
+
+        while (shard_queue.pop(block))
+        {
+            Stopwatch watch;
+            dictionary.blockToAttributes(block);
+            UInt64 elapsed_us = watch.elapsedMicroseconds();
+            if (elapsed_us > 1'000'000)
+                LOG_TRACE(dictionary.log, "Block processing for shard #{} is slow {}ms (rows {}).", shard, elapsed_us, block.rows());
+        }
+
+        if (!shard_queue.isFinished())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not pull non finished shards queue #{}", shard);
+    }
+
+    /// Split block to shards smaller block, using 'selector'.
+    void splitBlock(const IColumn::Selector & selector, const Block & block, Blocks & splitted_blocks)
+    {
+        size_t columns = block.columns();
+        for (size_t col = 0; col < columns; ++col)
+        {
+            MutableColumns splitted_columns = block.getByPosition(col).column->scatter(shards, selector);
```

Comment:
This `scatter` can work exteremely bad on real data with skewed distribution. Additionally it can allocate a lof of memory. Maybe we can just compute shard index column from hash `shard_index`, add fake column to block and then each thread will process block multiple times. But add element only if `shard_index` matches. That way we do not perform ```scatter```. But the problem is that we iterate data multiple times. Need to test this.

Reply:
>and then each thread will process block multiple times

This was my first try (note that, that version, does not calculate hash in `getShard` , so it was almost the same as you suggested), but it was slower, and besides with this approach it will work by the slowest shard in the dictionary, and there will be always someone that slower then others, due to rehash. And this is the reason why `ParallelDictionaryLoader` had been introduced, with it, shards scales loading of the dictionary almost linearly.

Reply:
I suspect that it scales almost linearly on your specific use case. If your data has skewed distribution it cannot be scaled linearly, and real data always almost always has skewed distribution.
I am not sure why this approach will work by the slowest shard in dictionary. For each shard you can maintain queue of blocks that it need to process, the same as right now. We can specify setting max_parallel_loaded_blocks that control how many blocks can be simualteneously in memory.
The main difference is that this method does not use `scatter`, which can be extremely unefficient.

Reply:
> I suspect that it scales almost linearly on your specific use case. If your data has skewed distribution it cannot be scaled linearly, and real data always almost always has skewed distribution.

If you have hashes then it will likely have normal distribution, but I see your point.

> I am not sure why this approach will work by the slowest shard in dictionary. For each shard you can maintain queue of blocks that it need to process, the same as right now. We can specify setting max_parallel_loaded_blocks that control how many blocks can be simualteneously in memory.

Do avoid lots of ping-pongs here, do I understand you correctly that you suggesting something like this - `blockToAttributes(block, shard_num)` - that loads only queries related to `shard_num` (to avoid locking)?

I had something like this in the original patch in fact, and it will be OK for me either I think.

Reply:
Yes. When we read block from dictionary storage. We compute additional `shard_hash` column. Then block can be added to each shard queue. Each shard processes data from its queue and use `blockToAttributes` with `shard_num`.
The downside is that we can have situation where we read 10 blocks and 9 shard process them sucesfully but some shard can't processes only 1 block for some reason because of data distribution. That way we can probably add setting, `max_parallel_blocks` that will control how many blocks we can keep in memory simultaneously. Such setting is probably easy to explain to client.
I am not sure if this will be faster than your approach with `scatter`, we need to test it.

Reply:
@kitaisreal I've tried this approach, but it does not scales as well as scatter + proper data re-distribution.

So now it works like this:
- getShard wraps the key with a hash function (`intHashCRC32`), this should give more or less normal distribution at the exit
- that way `scatter` should work OK
- and now `blockToAttributes` loads only small part of data that is related to particular shard (if any)

And I also removed `getShard` from `blockToAttributes`, since `shard` can be used, that is passed via argument.

And now it works almost as fast as it was (13.6% slower the previous version).

You can also find some details in the commit messages.